### DISCUSSION TITLE
Dagger SDK for Swift and challenges of macOS

### DISCUSSION THREAD

**Discussion by pepicrft** [dagger/dagger#8280] [General]
I think Dagger is amazing and I'd like to bring its awesomeness to the Apple development ecosystem, where there's a strong appetite from moving away from [Fastlane](https://github.com/fastlane/fastlane)'s Ruby approach to automation. Developers want to write their automation in Swift, and I think Dagger is the missing piece, not only to address that, but to start breaking the strong dependency with proprietary CI pipeline formats.

### Swift SDK
Following the SDK's `CONTRIBUTING.md` guidelines, I built a small prototype for a [Swift SDK](https://github.com/tuist/swift-sdk-for-dagger). The experience was a bit more involved than what I thought because I couldn't find any GraphQL query builder that I could reuse, but I got it working. I plan to write the logic for generating the code, and then open a PR in the https://github.com/dagger/dagger SDK. Any pointes regarding recommended next steps are appreciated.

### macOS-specific commands
A unique characteristic of automation for Apple platforms is that some tools are only available on macOS. This is a bit in conflict with Dagger's sandboxing model, which enable features such caching, so I was wondering if you'd be open to consider an escape hatch that developers can opt-in, either manually or automatically. Take for instance `xcodebuild`'s Xcode's CLI for compiling projects. I'd expect developers to do something like this in their Dagger pipelines:

```swift
let result = Dagger.connect().run(["xcodebuild", "-project", pathToProject])
```

Based on some conversations with @shykes, it seems that the team is keen to explore solutions to enable the above, so I'd like to know if I can support in making this happen. I can contribute the code changes myself, but I'd need some guidance on what aligns best with the interface and the principles of the tool.

### Officialising the SDK
While building the SDK, I wonder what it takes to make an SDK official. Based on the user-facing documentation, the impression that I got is that Dagger has some detection mechanism to know whether someone is using a particular SDK (e.g. Swift) and then it knows how to build and parse the pipeline or set of functions. My question is, is that done through static analysis of the code (e.g. for turning comments into the help of the CLI)? And if so, does that logic live inside Dagger's codebase?

I'm very excited about the huge opportunity this opens up for Dagger in the Apple development ecosystem. As the maintainer of [Tuist](https://github.com/tuist/tuist), we'd like to recommend Dagger as the official automation tool when creating projects, and I want to leverage our resources in making Dagger shine with Swift.

**Comment by helderco** [comment]
> [...] I plan to write the logic for generating the code, and then open a PR in the https://github.com/dagger/dagger SDK. Any pointes regarding recommended next steps are appreciated.

If I understand correctly, you've prototyped the client library and need pointers to start on codegen to generate the library from the API schema next?

Is there a library in Swift to make GraphQL requests? Does it support making an "introspection query" or do you need to do it "yourself"? To be clear, you don't have to. You can get the introspection (as JSON) from our `codegen` CLI in the dagger/dagger repo.

> While building the SDK, I wonder what it takes to make an SDK official. Based on the user-facing documentation, the impression that I got is that Dagger has some detection mechanism to know whether someone is using a particular SDK (e.g. Swift) and then it knows how to build and parse the pipeline or set of functions. My question is, is that done through static analysis of the code (e.g. for turning comments into the help of the CLI)? And if so, does that logic live inside Dagger's codebase?

There’s no “official” SDKs, only stable and experimental. It turns out that the current stable SDKs are the ones maintained by the Dagger team (Go, Python, and TypeScript), while all the experimental ones depend on community contributions.

However, am I correct in assuming that what you really want to know is how to enable running Dagger Modules and Functions with Swift?
```
❯ dagger init --sdk=swift my-module
❯ dagger functions -m my-module
```

If so, that part isn’t documented yet but basically you’ll need to add support for it in the SDK, and create a container runtime module that knows how to run Swift code. Then there’s a small addition to the Dagger Engine to allow using the `--sdk=swift` shorthand, but if not, you can actually use a remote git URL (or [local path](https://github.com/vito/daggerverse/blob/f6785759f2b2e79bc06ad85aef7c81893870e120/apko/dagger.json#L3)).

These are the current runtime modules in our repo:

- Python: https://github.com/dagger/dagger/tree/main/sdk/python/runtime
- TypeScript: https://github.com/dagger/dagger/tree/main/sdk/typescript/runtime
- PHP: https://github.com/dagger/dagger/tree/main/sdk/php/runtime
- Elixir: https://github.com/dagger/dagger/tree/main/sdk/elixir/runtime

While this isn’t yet all documented, you can ping me directly (same handle in Discord) so I can help you one step at a time.

**Comment by shykes** [comment]
Thank you for starting this conversation!

copying @grouville and @aduermael based on past conversations!

**Comment by pepicrft** [comment]
Hi folks!
Sorry for my late response, I've been busy with other stuff.

> f I understand correctly, you've prototyped the client library and need pointers to start on codegen to generate the library from the API schema next?
> Is there a library in Swift to make GraphQL requests? Does it support making an "introspection query" or do you need to do it "yourself"? To be clear, you don't have to. You can get the introspection (as JSON) from our codegen CLI in the dagger/dagger repo.

I couldn't find a GraphQL client in Swift with a builder-like interface to construct a query, so I built a simple thing myself for the sake of the prototype. I'm thinking about using Dagger's codegen CLI to generate the Swift code then. The alternative would be to use Swift macros, which can be used to generate Swift code at compilation-time. The upside of doing that is that you don't need to release new versions of the SDK when the schema changes. The downside is that it'll increase the code compilation that depends on the SDK (i.e. pipelines) significantly. What would you advise?

> If so, that part isn’t documented yet, but basically, you’ll need to add support for it in the SDK and create a container runtime module that knows how to run Swift code. Then there’s a small addition to the Dagger Engine to allow using the --sdk=swift shorthand, but if not, you can actually use a remote git URL (or [local path](https://github.com/vito/daggerverse/blob/f6785759f2b2e79bc06ad85aef7c81893870e120/apko/dagger.json#L3)).

Thanks for the information. So it means that the SDK is external to Dagger (either officially or non-officially supported), but there's a part of the "supporting an SDK", which lives in Dagger itself. Did I get it right?

> While this isn’t yet all documented, you can ping me directly (same handle in Discord) so I can help you one step at a time.

I'll do so, thanks!

### CLI help
Out of curiosity, how does Dagger extract the functions, arguments, and the documentation from pipelines written in multiple languages? I assume it does it through static analysis of the code, but I want to double-check. If so, is this code that also lives in Dagger itself? Thanks!

### Next steps
- I'll continue with the code generation and update you here.
- I'll look into creating a container runtime for Swift

It'd be awesome if we could find a way for the engine to run processes in the host. I believe this would be the only piece to evangelize the Dagger Swift SDK in the Swift community once it's complete. I'd like to leverage the presence that https://tuist.io has in that community to spread the goodness of this technology.

**Reply by helderco** [reply]
> I couldn't find a GraphQL client in Swift with a builder-like interface to construct a query, so I built a simple thing myself for the sake of the prototype.

The Go SDK also has a custom query builder (and TypeScript SDK too). You can get inspiration here: https://github.com/dagger/dagger/tree/main/sdk/go/querybuilder

One thing to note is that the query builder doesn’t have to support every feature a usual third-party GraphQL query builder would, because we use a simpler single field chain.

There’s some available documentation on building a client. See [SDK Contribution Guide > Dagger client](https://github.com/dagger/dagger/blob/main/sdk/CONTRIBUTING.md#dagger-client).

---

> I'm thinking about using Dagger's codegen CLI to generate the Swift code then. The alternative would be to use Swift macros, which can be used to generate Swift code at compilation-time. The upside of doing that is that you don't need to release new versions of the SDK when the schema changes. The downside is that it'll increase the code compilation that depends on the SDK (i.e. pipelines) significantly. What would you advise?

 Dagger’s [`codegen`](https://github.com/dagger/dagger/tree/main/cmd/codegen/generator) CLI can be used to just get the introspection result in JSON instead of doing it all, and use that as input for a Swift based codegen. The advantage is that you don’t need to do any GraphQL calls in this case, but this is only really needed when publishing a library package for use **without modules**, because the engine itself already provides this JSON file to the SDK's module for running codegen.

However, I suggest using that CLI for generating the client bindings as well (the whole thing) since I think we’ll eventually require it. The only thing that gives me pause is that the current Go based codegen implementation needs quite a bit of refactoring in order to be easier to use and manage.

See:
- https://github.com/dagger/dagger/issues/5226

---

> > [...] there’s a small addition to the Dagger Engine to allow using the --sdk=swift shorthand, but if not, you can actually use a remote git URL (or [local path](https://github.com/vito/daggerverse/blob/f6785759f2b2e79bc06ad85aef7c81893870e120/apko/dagger.json#L3)).
> 
> Thanks for the information. So it means that the SDK is external to Dagger (either officially or non-officially supported), but there's a part of the "supporting an SDK", which lives in Dagger itself. Did I get it right?

Yes. So, without touching core, you can do this:

```
dagger init --sdk=github.com/pepicrft/dagger-swift-sdk my-module
```

Although we [ask you](https://github.com/dagger/dagger/blob/main/sdk/CONTRIBUTING.md#how-do-i-contribute-a-new-sdk) to make a PR request against `sdk/swift`:

```
dagger init --sdk=github.com/dagger/dagger/sdk/swift my-module
```

Which can be developed using:

```
cd sdk/swift
dagger init --sdk=. my-module
```

What can be added in core is support for an alias, which is [trivial to add](https://github.com/dagger/dagger/blob/36cfba5b1cd75c7b672471167f0dee4b9942788c/core/schema/sdk.go#L175-L178):

```
dagger init --sdk=swift my-module
```
That will point to the latest release by default, but users can specify a specific version (or branch).

---

> ### CLI help
> Out of curiosity, how does Dagger extract the functions, arguments, and the documentation from pipelines written in multiple languages? I assume it does it through static analysis of the code, but I want to double-check. If so, is this code that also lives in Dagger itself? Thanks!

The CLI makes a single [GraphQL query to introspect](https://github.com/dagger/dagger/blob/36cfba5b1cd75c7b672471167f0dee4b9942788c/cmd/dagger/typedefs.graphql) what’s on the current API schema. It uses this information in order to build the commands and their flags.

But, it’s each SDK’s responsibility to analyze the actual source code and report to Dagger what the module’s available objects and functions are. This is all done via API calls, both for the SDK to know if it should register the functions, or invoke a specific one, but also to return results.

The way this is hooked up is you need to create a Go module (conventionally located in `sdk/<sdk>/runtime`) that creates a `Container` that is able to run your SDK’s code (i.e., Swift). That container will need to set an entrypoint, which is what the engine executes anytime it needs to either get the available functions or invoke a function call.

The entrypoint doesn’t get any arguments, it’s the entrypoint’s responsibility to check with an API call which one is needed.

**Reply by pepicrft** [reply]
Thanks a lot, @helderco, for your answers 🙏🏼. That's all I needed to continue my work. It'll take me a bit of time because I'm distracted by something else, but hopefully, I'll have something in the following weeks.

As mentioned above, we'll need to run steps in the host because some tools are macOS-only, for example, `xcodebuild`, which Xcode developers use to compile their apps. Has the Dagger team discussed or planned any work around solving that need?
